control-plane/agents/src/bin/core/controller/resources/operations.rs

```diff
@@ -1,6 +1,64 @@
 use crate::controller::registry::Registry;
 use agents::errors::SvcError;
 
+/// Resource Cordon Operations.
+#[async_trait::async_trait]
+pub(crate) trait ResourceCordon {
+    type CordonOutput: Sync + Send + Sized;
+    type UncordonOutput: Sync + Send + Sized;
+
+    /// Cordon the resource.
+    async fn cordon(
+        &mut self,
+        registry: &Registry,
+        label: String,
+    ) -> Result<Self::CordonOutput, SvcError>;
+    /// Uncordon the resource.
+    async fn uncordon(
+        &mut self,
+        registry: &Registry,
+        label: String,
+    ) -> Result<Self::UncordonOutput, SvcError>;
+}
+
+/// Resource Drain Operations.
+#[async_trait::async_trait]
+pub(crate) trait ResourceDrain {
+    type DrainOutput: Sync + Send + Sized;
+    /// Drain the resource.
+    async fn drain(
+        &mut self,
+        registry: &Registry,
+        label: String,
+    ) -> Result<Self::DrainOutput, SvcError>;
+    /// Mark the resource as drained.
+    async fn set_drained(&mut self, registry: &Registry) -> Result<Self::DrainOutput, SvcError>;
+}
+
+/// Node drain Operations.
+#[async_trait::async_trait]
+pub(crate) trait ResourceNodeDrain {
```

Comment:
Is this trait needed or could we just implement the methods against Guard<NodeSpec> ?

Reply:
I was wondering about that. I suppose I was following the existing pattern too closely. Now fixed.